src/realm/node_header.hpp

```diff
@@ -229,16 +301,631 @@ class NodeHeader {
+{
+    // TODO optimize
+    NodeHeader::set_wtype_in_header(wtype_Bits, header);
+    NodeHeader::set_size_in_header(num_elements, header);
+}
+template <>
+inline void NodeHeader::set_num_elements<NodeHeader::Encoding::WTypMult>(char* header, size_t num_elements)
+{
+    // TODO optimize
+    NodeHeader::set_wtype_in_header(wtype_Multiply, header);
+    NodeHeader::set_size_in_header(num_elements, header);
+}
+template <>
+inline void NodeHeader::set_num_elements<NodeHeader::Encoding::WTypIgn>(char* header, size_t num_elements)
+{
+    // TODO optimize
+    NodeHeader::set_wtype_in_header(wtype_Ignore, header);
+    NodeHeader::set_size_in_header(num_elements, header);
+}
+template <>
+inline void NodeHeader::set_num_elements<NodeHeader::Encoding::AofP>(char* header, size_t num_elements)
+{
+    REALM_ASSERT_DEBUG(get_encoding(header) == Encoding::AofP);
+    REALM_ASSERT_DEBUG(num_elements < 0x10000);
+    (reinterpret_cast<uint16_t*>(header))[3] = static_cast<uint16_t>(num_elements);
+}
+template <>
+inline void NodeHeader::set_num_elements<NodeHeader::Encoding::PofA>(char* header, size_t num_elements)
+{
+    REALM_ASSERT_DEBUG(get_encoding(header) == Encoding::PofA);
+    REALM_ASSERT_DEBUG(num_elements < 0x10000);
+    (reinterpret_cast<uint16_t*>(header))[3] = static_cast<uint16_t>(num_elements);
+}
+
+
+template <>
+inline void NodeHeader::set_arrayA_num_elements<NodeHeader::Encoding::Flex>(char* header, size_t num_elements)
+{
+    REALM_ASSERT_DEBUG(get_encoding(header) == Encoding::Flex);
+    REALM_ASSERT_DEBUG(num_elements < 0b10000000000); // 10 bits
+    uint16_t word = (reinterpret_cast<uint16_t*>(header))[3];
+    word &= ~(0b1111111111 << 10);
+    word |= num_elements << 10;
+    (reinterpret_cast<uint16_t*>(header))[3] = word;
+}
+template <>
+inline void NodeHeader::set_arrayB_num_elements<NodeHeader::Encoding::Flex>(char* header, size_t num_elements)
+{
+    REALM_ASSERT_DEBUG(get_encoding(header) == Encoding::Flex);
+    REALM_ASSERT_DEBUG(num_elements < 0b10000000000); // 10 bits
+    uint16_t word = (reinterpret_cast<uint16_t*>(header))[1];
+    word &= ~(0b1111111111 << 10);
+    word |= num_elements << 10;
+    (reinterpret_cast<uint16_t*>(header))[1] = word;
+}
+
+template <>
+inline size_t NodeHeader::get_num_elements<NodeHeader::Encoding::Packed>(const char* header)
+{
+    REALM_ASSERT_DEBUG(get_encoding(header) == Encoding::Packed);
+    return (reinterpret_cast<const uint16_t*>(header))[3];
+}
+template <>
+inline size_t NodeHeader::get_num_elements<NodeHeader::Encoding::AofP>(const char* header)
+{
+    REALM_ASSERT_DEBUG(get_encoding(header) == Encoding::AofP);
+    return (reinterpret_cast<const uint16_t*>(header))[3];
+}
+template <>
+inline size_t NodeHeader::get_num_elements<NodeHeader::Encoding::PofA>(const char* header)
+{
+    REALM_ASSERT_DEBUG(get_encoding(header) == Encoding::PofA);
+    return (reinterpret_cast<const uint16_t*>(header))[3];
+}
+template <>
```

Comment:
I don't like having 3 different functions that actually are the same. What is the purpose? It does not increase readability? The same goes for other functions. Perhaps the templating has taken over here :-)

Reply:
Ok I will revert the code to have one single method and pass the encoding as argument.